remove sigmoid optimization for extreme values

bridgeland opened the pull request:
the sigmoid optimization was inaccurate for z > 19

dfm commented on theano/tensor/nnet/sigm.py:
```diff
@@ -81,12 +81,12 @@ def c_code(self, node, name, inp, out, sub):
         ):
             return (
-                """%(z)s = %(x)s < -88.0f ? 0.0 : %(x)s > 15.0f ? 1.0f : 1.0f /(1.0f + exp(-%(x)s));"""
+                """%(z)s = 1.0f /(1.0f + exp(-%(x)s));"""
```
The style unit test are failing and the fix should be as easy as changing this line to

```
            return """%(z)s = 1.0f /(1.0f + exp(-%(x)s));""" % locals()
```

instead of the multi-line version to be consistent with [black](https://github.com/psf/black).

dfm commented on theano/tensor/nnet/sigm.py:
```diff
@@ -81,12 +81,12 @@ def c_code(self, node, name, inp, out, sub):
         elif node.inputs[0].type == scalar.float64:
             return (
-                """%(z)s = %(x)s < -709.0 ? 0.0 : %(x)s > 19.0 ? 1.0 : 1.0 /(1.0+exp(-%(x)s));"""
+                """%(z)s = 1.0 /(1.0+exp(-%(x)s));"""
```
Similarly, this one should be:
```
            return """%(z)s = 1.0 /(1.0+exp(-%(x)s));""" % locals()
```
to get the tests to pass.

bridgeland commented:
@dfm : thanks. Done

dfm commented:
@bridgeland: Thanks! Unfortunately that isn't quite right, the string needs to be on the same line as the return (`return """...` - they do call black "uncompromising" after all :/). I know this is annoying and I might be able to push to this branch directly to debug it later today, if you'd prefer! But, I think that after changing that, we should be able to get the other tests running. Thanks and sorry about the hoops!

bridgeland commented:
Wups! 

Fixed.

brandonwillard commented:
To fix these kinds of formatting issues, you should only need to run `black theano/ tests/ setup.py`.   

Otherwise, these commits need to be squashed/rebased, and this change needs a test that highlights/demonstrates the bug that it's addressing.

bridgeland commented:
@brandonwillard I squashed and rebased. But I cannot create a test, for two reasons. 
First I was not able to get the existing theano tests to pass: 654 errors, 28 failures. 
Obviously there is some problem with my installation, perhaps with Xcode. 

Second, I am not sure how to write an automated test for this change. My original [hand test](https://github.com/pymc-devs/Theano-PyMC/issues/29) still exhibits ~15 divergences (versus ~600 previously), since
occasionally a value is generated that exceeds the now greater limits.

brandonwillard commented:
From the looks of it, you could add/modify a test to confirm that the operator computes values outside of the previous cut-offs.  The existing tests are [here](https://github.com/pymc-devs/Theano-PyMC/blob/master/tests/tensor/nnet/test_sigm.py#L33), so that's where any changes and additions would go.

I don't know much about Xcode, but I don't think it should be involved.  If you create a clean Conda environment and install install the requirements (see [here](https://github.com/pymc-devs/Theano-PyMC/blob/master/.travis.yml#L62)), then the tests should run fine from the command line (e.g. `pytest tests`).  If they don't, then that's a separate problem; one that would be worth addressing on its own.

Also, since there are a lot of tests, it's often more efficient to simply run the new/altered test(s) and any others that might be affected by your changes.  You can specify very specific tests&mdash;or a single test module/file&mdash;in the `pytest` command line options.  After confirming that your new test(s) pass&mdash;and that related ones don't break&mdash;locally, you can push the changes to this branch and let CI perform the other tests.

bridgeland commented:
@brandonwillard Hmm. Perhaps my mistake was to run tests using **nosetest tests** instead of **pytest tests**.

I am still getting some failures, and a lot of skipped tests.

![image](https://user-images.githubusercontent.com/5245697/92047502-4857bf00-ed53-11ea-837e-1c4ff60b576e.png)

Is this expected? Also should it take 90 minutes to run all the tests?

brandonwillard commented:
It looks like everything passed in that output.  If you're referring to `6 xfailed`, those are _expected_ failures.

Unfortunately, 90 minutes to run all of the tests is not surprising.  That's why I recommend running only the relevant test modules.

bridgeland commented:
I am stuck. As you suggested, **tests/tensor/nnet/test_sigm.py** --- or its imports --- 
would seem to be a good place to put a new test of this situation. So I tried this 
addition, in **tests/tensor/test_basic.py**:

![image](https://user-images.githubusercontent.com/5245697/92176525-9da3d700-ee0c-11ea-94d3-524b341980bc.png)

That caused the expected failure, without the **c_code()** fix in place:

![image](https://user-images.githubusercontent.com/5245697/92175357-7d731880-ee0a-11ea-992b-898fbd27b3ed.png)

Sadly the same failure occurs with the fix in place. A bit of logging within
**ScalarSigmoid.c_code()** reveals that the fixed function is never invoked
in the new test case, and in fact is only invoked three times in  

     $ pytest -s tests/tensor/nnet/test_sigm.py::TestSigmoidBroadcast

corresponding to these integer test cases:

![image](https://user-images.githubusercontent.com/5245697/92176471-8533bc80-ee0c-11ea-9103-8bf5e3aa834e.png)

So I am at a loss. Despite much experimentation, I cannot write a test case
that exercises this fix.

Summary:
1. The fix works, as demonstrated by the changed behavior of the pymc3 beta 
non-automated test
2. My attempted test case does not  exercise the fix, as **ScalarSigmoid.c_code()**  
is never invoked.
3. I don't understand the conditions under which **ScalarSigmoid.c_code()**  is 
invoked.

bridgeland commented:
A bit more explanation. 

The correct test to modify seems to be here, in **tests/tensor/nnet/test_sigm.py**:

![image](https://user-images.githubusercontent.com/5245697/92239254-ae922e00-ee88-11ea-9d93-003aa78fd118.png)

That test uses the dict **_good_broadcast_unary_normal_no_complex**, imported from
**tests/tensor/test_basic.py**, which lists a bunch of different cases to test:

![image](https://user-images.githubusercontent.com/5245697/92239372-e305ea00-ee88-11ea-9b79-c2de8353146f.png)

Experimentation reveals that the cases that test ints generate C code,
and the cases that test floats do not. 

Why? Is there a way to change that, to force C to be emitted?

@dfm @brandonwillard : do either of you understand the theano code generation?

brandonwillard commented:
@bridgeland, there is a code/compilation caching mechanism that might be affecting your measurements.  For instance, I don't think `ScalarSigma.c_code` will be called twice if/when the cache is used; however, the C code/linker will still be used at runtime (if compilation was successful).

You can check whether or not `theano.config.cxx` is non-empty, because&mdash;if it is empty&mdash;the C code won't be used.  Otherwise, I don't see any parameters in `makeBroadcastTester` and `makeTester` that specifically set the C/Python linker, so the remaining options that would affect the choice between the C/Python operator code are most likely environmental (e.g. environment variables like `FAST_COMPILE` and others used in [the Travis setup](https://github.com/pymc-devs/Theano-PyMC/blob/master/.travis.yml#L76)).

bridgeland commented:
@brandonwillard  A cached compilation would explain the behavior I am seeing, but 
only if the cache is persisted in the file system. A cache in memory would presumably
be cleared with a new invocation of **pytest**. **pytest** is exhibiting  no calls to 
`ScalarSigmoid.c_code()` for any float tests, which would seem to be incompatible
with an in-memory cache. 

`theano.config.cxx` does not seem to be involved, as it has a value of `'/usr/bin/clang++'`
both for the int cases where `c_code() `is called and the float cases where it is not called.

bridgeland commented:
Aha! Clearing the cache solves all my problems:

>     $ theano-cache clear

(Well not all my problems. For example, the DC music clubs are still closed due to the
virus. But at least this theano problem is solved.)

brandonwillard commented:
> (Well not all my problems. For example, the DC music clubs are still closed due to the
> virus. But at least this theano problem is solved.)

By solving Theano problems we can solve PyMC3 problems, and by solving PyMC3 problems we can solve modeling problems, and by solving modeling problems, we can solve virus problems...

bridgeland commented:
Arg. It looks like my new test case failed lint.

bridgeland commented:
Hmm. I fixed the earlier lint issues: unwanted whitespace. 

Now I have new lint issues, but I can't find the description of
what is wrong.

twiecki commented on theano/tensor/nnet/sigm.py:
```diff
@@ -57,41 +57,6 @@ def grad(self, inp, grads):
 
         return [rval]
 
-    def c_code(self, node, name, inp, out, sub):
```
Is this a test?

bridgeland replied:
No. I added a case to the existing tests with the changes to **tests/tensor/test_basic.py** and **tests/tensor/nnet/test_sigm.py**.

The change to **theano/tensor/nnet/sigm.py** was to completely delete the errant optimization.

twiecki replied:
I see, thanks.

brandonwillard replied:
I don't understand; the C implementation shouldn't be removed.

twiecki commented:
Did you run black?
```
black -t py36 --check theano/ tests/ setup.py

would reformat /home/travis/build/pymc-devs/Theano-PyMC/tests/tensor/nnet/test_sigm.py

would reformat /home/travis/build/pymc-devs/Theano-PyMC/tests/tensor/test_basic.py
```

bridgeland commented:
@twiecki Just ran black and updated two files in the fix.

Note to self: run black as part of my normal python dev process

bridgeland commented:
Hmm. It seems that at least of the tests — **tests/theano/tensor/nnet/test_nnet.py::test_argmax_pushdown**
— depends on the existence of the deleted `ScalarSigmoid.c_code()`: it fails when that method is removed
but passes when that method is present even if it does nothing.

The test does not look sound:

```
        # print 'AFTER'
        # for node in fgraph.toposort():
        # print node.op
        assert len(fgraph.toposort()) == 3    # bridgeland: why 3?
        assert isinstance(fgraph.toposort()[0].op, tensor.Elemwise)
        assert isinstance(fgraph.toposort()[1].op, Softmax)
        assert isinstance(fgraph.toposort()[2].op, tensor.CAReduce)
        assert isinstance(fgraph.toposort()[2].op.scalar_op, theano.scalar.Maximum)
```

But I am reluctant to change the test as I do not really understand it.

Will try again with a noop `ScalarSigmoid.c_code()`

brandonwillard commented:
I changed this PR into a draft, just in case you were testing changes via the CI.

bridgeland commented:
This time I ran all the tests locally before pushing, not just the 35 tests in 
**tests/tensor/nnet/test_sigm.py**. (They all pass.)

@brandonwillard I appreciate that deleting the C implementation of `SigmoidScalar` causes
some tests to fail. But why should anything fail, given that the C implementation is identical
to what is calculated without it? Either some tests are unnecessarily fussy, or (more likely)
I am misunderstanding something.

twiecki commented:
without a c implementation it will be very slow.

On Mon, Sep 7, 2020, 13:32 David M Bridgeland <notifications@github.com>
wrote:

> This time I ran all the tests locally before pushing, not just the 35
> tests in
> *tests/tensor/nnet/test_sigm.py*. (They all pass.)
>
> @brandonwillard <https://github.com/brandonwillard> I appreciate that
> deleting the C implementation of SigmoidScalar causes
> some tests to fail. But why should anything fail, given that the C
> implementation is identical
> to what is calculated without it? Either some tests are unnecessarily
> fussy, or (more likely)
> I am misunderstanding something.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/pymc-devs/Theano-PyMC/pull/32#issuecomment-688266043>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFETGE63K2GW4YWBB6JKODSETAFDANCNFSM4QOPFIYQ>
> .
>

bridgeland commented:
I am down to one failed test in the CI: **tests/tensor/nnet/test_sigm.py::TestSigmoidBroadcast::test_good**

That test passes locally in my test environment

![image](https://user-images.githubusercontent.com/5245697/92388482-5441d900-f0e5-11ea-9ffe-2b5881e98d58.png)

Not sure why it fails in the CI.

bridgeland commented:
Note that the error message for the failed test does not help us much:

![image](https://user-images.githubusercontent.com/5245697/92420156-cb0bc000-f13f-11ea-9ee7-82aa3c8938ee.png)

I suspect that the expected and the got (actual) differ in some small way
that rounds up to 1.0 when displayed. Is there a way to do 

```
     numpy.set_printoptions(precision=20)
```

for the CI?

bridgeland commented:
But the larger question is: why would this test pass in my local environment 
but fail in the CI?

Is there a way to determine the version of packages loaded in the CI?

bridgeland commented:
Does the CI clear the theano cache before running the tests?

bridgeland commented:
Here is the result of `$ conda list` for my local environment, the environment
in which all the tests pass:

```
# packages in environment at /Users/bridgeland/miniconda3/envs/theano_dev37:
#
# Name                    Version                   Build  Channel
attrs                     20.1.0                     py_0  
blas                      1.0                         mkl  
ca-certificates           2020.7.22                     0  
certifi                   2020.6.20                py37_0  
importlib-metadata        1.7.0                    py37_0  
importlib_metadata        1.7.0                         0  
iniconfig                 1.0.1                      py_0  
intel-openmp              2019.4                      233  
libcxx                    10.0.0                        1  
libedit                   3.1.20191231         h1de35cc_1  
libffi                    3.3                  hb1e8313_2  
libgfortran               3.0.1                h93005f0_2  
libgpuarray               0.7.6                h1de35cc_0  
mkl                       2019.4                      233  
mkl-service               2.3.0            py37hfbe908c_0  
mkl_fft                   1.1.0            py37hc64f4ea_0  
mkl_random                1.1.1            py37h959d312_0  
more-itertools            8.5.0                      py_0  
ncurses                   6.2                  h0a44026_1  
nose                      1.3.7                    py37_2  
numpy                     1.19.1           py37h3b9f5b6_0  
numpy-base                1.19.1           py37hcfb5961_0  
openssl                   1.1.1g               h1de35cc_0  
packaging                 20.4                       py_0  
parameterized             0.7.4                      py_0  
pip                       20.2.2                   py37_0  
pluggy                    0.13.1                   py37_0  
py                        1.9.0                      py_0  
pyparsing                 2.4.7                      py_0  
pytest                    6.0.1                    py37_0  
python                    3.7.9                h26836e1_0  
readline                  8.0                  h1de35cc_0  
scipy                     1.5.2            py37h912ce22_0  
setuptools                49.6.0                   py37_0  
six                       1.15.0                     py_0  
sqlite                    3.33.0               hffcf06c_0  
theano-pymc               1.0.5+12.gf2a0ad518           dev_0    <develop>
tk                        8.6.10               hb0a8c7a_0  
toml                      0.10.1                     py_0  
wheel                     0.35.1                     py_0  
xz                        5.2.5                h1de35cc_0  
zipp                      3.1.0                      py_0  
zlib                      1.2.11               h1de35cc_3 
```

bridgeland commented:
> But the larger question is: why would this test pass in my local environment
> but fail in the CI?
> 
> Is there a way to determine the version of packages loaded in the CI?

Never mind. I found it in the Travis results.

bridgeland commented:
I attempted to recreate the exact environment in the failing CI test, using the 
commands in the failing CI test. The attempt was mostly successful—the
environments mostly match, with the exceptions listed below. But despite 
recreating the environment (mostly), I could not replicate the failure. So I 
suspect one of these environment differences:

1. A somewhat older version of **mkl** is installed in my local environment
than in the CI: 2019.4 versus 2020.2. When I attempt to install version 
2020.2 explicitly, conda fails, reporting many conflicts.

2. A somewhat older version of **intel-openmp** is installed in my local
environment than in the CI: 2019.4 versus 2020.2. When I attempt
to install the new version explicitly, conda fails, reporting many
conflicts.

3. There are several packages that are installed in the CI but not
installed in my local environment. These packages look like either
linux dependencies or gcc dependencies, not needed for running 
on macos: **_libgcc_mutex**, **ld_impl_linux-64**, **libgcc-ng**, 
**libtool**, **libxcb**.

4. The fortran library installed in my local environment is completely
different from the one installed in the CI: **libgfortran** (v3.0.1) versus
**libgfortran-ng** (v7.3.0)

5. The C++ library installed in my local environment is different from
the one installed in the CI: **libcxx** (v1.16) vs. **libstdcxx-ng** (v7.3.0).
Seems unlikely to be the problem.

6. The text conversion library **libiconv** (v1.16) is installed in my 
local environment but not in the CI. Seems unlikely to be the problem.

7.  My local environment has the package **gettext** installed (v 0.19.8.1),
but the CI does not. **gettext** was installed as a dependency of 
graphviz. Seems unlikely to be the problem.

bridgeland commented:
A clue: the tests with PART="tests/tensor/nnet --ignore-glob='*/test_abstract_conv.py'"
has a failure in Python 3.6, but the same PART does not fail in Python 3.7.

bridgeland commented:
CI job 126.12 fails on **tests/tensor/nnet/test_sigm.py::TestSigmoidBroadcast::test_good**
CI job 126.6 has no failing tests. The two jobs run the same tests, and have almost identical
environments, differing only in two ways:

1. Job 126.12 uses python 3.6.10 while job 126.6 uses python 3.7.9

2. The environment for job 126.12 includes the package **dataclasses**, while job 126.6 
does not. **dataclasses** is a backport to python 3.6 of the new python 3.7 feature for data
classes. (Note to self: useful new feature.)

Neither environmental difference seems likely to be the cause of the failure/non-failure
difference. So I no longer suspect an environmental difference.

Instead I am mystified. I cannot replicate the bug even when I (almost) replicate the environment
in macos. And I don't see any difference between job 126.12 and job 126.6 that would cause
the former to fail but not the latter.

Short of standing up my own local linux box and experimenting, I don't know where to go next.

@brandonwillard @twiecki @dfm Help?

brandonwillard commented:
@bridgeland, are you running this test locally under the same configuration as CI (e.g. with the same `FAST_COMPILE` and `FLOAT32` values)?

bridgeland commented:
> @bridgeland, are you running this test locally under the same configuration as CI (e.g. with the same `FAST_COMPILE` and `FLOAT32` values)?

Not sure. $FLOAT32 and $FAST_COMPILE are both blank in my local config. Where do I see what they are in
job 126.12?

brandonwillard commented:
These lines in the [`.travis.yml`](https://github.com/pymc-devs/Theano-PyMC/blob/master/.travis.yml#L76) determine the CI test config, and they're applied to the CI environment [here](https://travis-ci.org/github/pymc-devs/Theano-PyMC/jobs/725056984#L406).

Looking at that failure, CI appears to be using `float64`&mdash;so make sure you're using that setting locally&mdash;and the computed/expected results differ somewhere near the `1e-9` range, while the actual `eps` used for comparison is probably set to `1e-10`.  Other environment variables are set there, as well, and those&mdash;along with broader system-specific conditions and version differences&mdash;could affect Python implementation and C compilation choices such that both end up having slightly different ranges of significant digits (e.g. changes in the order of operations and/or precision of intermediate steps).  See [here](https://gcc.gnu.org/wiki/FloatingPointMath) for some relevant settings on the C side.

I would first try to determine whether or not `1e-10` is actually within the range of significant digits for this function, precision, and test values.  If you can determine that it isn't, then the fix is easy: lower the `eps` tolerance.  Otherwise, if you can determine that it is, then we need to find out which implementation is producing the correct result and fix the other one (or fix the test framework/utilities).

bridgeland commented:
@brandonwillard I missed that. Will investigate.

bridgeland commented:
$FLOAT32 and $FAST_COMPILE are not the problem. They are set to 1—in various combinations—
in the Python 3.7 CI tests, but remain unset in the Python 3.6 CI tests. 

![image](https://user-images.githubusercontent.com/5245697/92644033-b3a01480-f2b0-11ea-8edd-0524ffa0ee23.png)

(Although looking at the travis.yml, I don't see how that specification indicates
to set $FLOAT32 and $FAST_COMPILE for Python 3.7 and not for Python 3.6,
unless `normaltest` means for the first python (3.7) and for no others.

![image](https://user-images.githubusercontent.com/5245697/92645519-f6fb8280-f2b2-11ea-9369-92a64889bd11.png)

float64 is in fact set locally, or at least that is my interpretation of the results of 
`print(theano.config.__str__(print_doc=False))`:

![image](https://user-images.githubusercontent.com/5245697/92644459-63758200-f2b1-11ea-87df-978b059cfbfe.png)

I'll investigate `eps`.

bridgeland commented:
Just pushed a branch that only increases the numpy print precision,
to investigate eps. Will remove once we better understand.

bridgeland commented:
Wow! Lint failed (note to self), but this test build passes all tests, even
though the only change was adding the line

```
                np.set_printoptions(precision=20)
```

to print out the error values with more precision. 

I do not understand, but I will try again, dropping that line and see
if that caused the changed behavior.

bridgeland commented:
Hmm. The CI passed now. Not clear why it passed now and failed before. 
Does it matter?

The test coverage has decreased a bit. This is surprising since the only
lines I changed are still covered, and the changes to the tests add a case
but don't mess with existing cases. 

@brandonwillard Any explanation?

bridgeland commented:
For example, in the file **theano/tensor/nnet/sigm.py**, there are many lines
that are not covered by tests, and marked in red. A few of those lines have a
little red UNCOV marking on the left, which I take to mean that they are newly
uncovered by this change.  For example, there are four such uncovered lines in 
`UltraFastScalarSigmoid.c_code()`, and in fact that method is entirely uncovered:

![image](https://user-images.githubusercontent.com/5245697/92940851-9ccb0080-f41d-11ea-8fec-3ab74e98d42d.png)

How did this method get uncovered by my change? I have no idea.
Neither the small fix I made nor the additional test case affects UltraFastScalarSigmoid
AFAIK. It would seem to be unaffected by the few changes I made.

Perhaps I am misreading the results, and there is some other test coverage
problem that I am missing.

bridgeland commented:
@brandonwillard @dfm @twiecki @aseyboldt Any ideas on what would account for this
decrease in test coverage? Any suggestions for how to investigate?

FWIW, I ran one of the test jobs locally (130.4), and examined the test coverage that
resulted from that test run. The results are somewhat different from that reported by
Travis CI. For example, for source file **theano/tensor/nnet/sigm.py** on this job, both
the local test and CI miss most statements. (Presumably this source file is better 
covered by other test configurations.) But there are 11 lines that are somehow
covered by the test locally and not covered by the CI running the same configuration.

bridgeland commented:
Or maybe the decline in test coverage does not matter? Should I mark as 
ready for review?

twiecki commented:
@bridgeland Yes, that doesn't matter, there's always noise with code coverage. This looks good to me but I'll wait for @brandonwillard and potentially @dfm to merge. Thanks for the contribution and sorry for all the troubles you ran into.

brandonwillard commented on theano/tensor/nnet/sigm.py:
```diff
@@ -80,15 +80,9 @@ def c_code(self, node, name, inp, out, sub):
-                """%(z)s = %(x)s < -709.0 ? 0.0 : %(x)s > 19.0 ? 1.0 : 1.0 /(1.0+exp(-%(x)s));"""
-                % locals()
-            )
+            return """%(z)s = 1.0 /(1.0+exp(-%(x)s));""" % locals()
```
Mind adding some spacing around the `1.0+exp` in the latter C string and an extra space after the `/(1.0` in both?  This isn't an important/functional change, but it is the best time to make such changes.

bridgeland replied:
done

brandonwillard reviewed (changes requested):
Squash these commits/rebase and we're good to merge.  We might be able to perform a squash merge from our end, but, with those merge commits, I'm not entirely sure.
